### PR TITLE
Fixed #25847 - Make User.is_(anonymous|authenticated) properties

### PR DESCRIPTION
**Pull request by jlaine**
Refactored User.is_anonymous and User.is_authenticated to make them
read-only properties instead of methods. This avoids making mistakes by
accidentally checking:

    request.user.is_authenticated

instead of:

    request.user.is_authenticated()

In order to maintain backwards compatibility, the properties' value is
a "callable boolean" as suggested by Aymeric Augustin: it behaves like
a boolean in a boolean context, but is also a callable.

### CODE REVIEW

File: django/utils/functional.py [hunk 1]
```diff
@@ -440,3 +440,30 @@ def partition(predicate, values):
...
+    """
+    def __init__(self, value):
+        self.value = value
+
+    def __bool__(self):
+        return self.value
+
+    def __call__(self):
+        warnings.warn(
+            "Using user.is_authenticated() and user.is_anonymous() as a method is deprecated.",
```
Comment (sergeykolosov) [inline django/utils/functional.py]: Does this imply `CallableBool` is intended to be used nowhere but with `is_authenticated` and `is_anonymous`?

Comment (jlaine) [inline django/utils/functional.py]: So far those are indeed the only usecases, so you are right: it feels as though it should be moved to django.contribute.auth. That is was I did originally but my problem was: "how do I get the template engine to not emit a deprecation warning?" Importing CallableBool from django.contribute.auth led to an import loop. Suggestions?

Comment (timgraham) [inline django/utils/functional.py]: Probably `utils/deprecation.py` would be an okay place.

File: tests/auth_tests/test_basic.py [hunk 2]
```diff
@@ -44,11 +45,18 @@ def test_user(self):
...
-        self.assertTrue(u.is_authenticated())
+        self.assertFalse(u.is_anonymous)
+        self.assertTrue(u.is_authenticated)
         self.assertFalse(u.is_staff)
         self.assertTrue(u.is_active)
         self.assertFalse(u.is_superuser)
 
+        # Backwards-compatibility callables
+        with self.assertRaises(RemovedInDjango20Warning):
+            self.assertFalse(u.is_anonymous())
```
Comment (timgraham) [inline tests/auth_tests/test_basic.py]: The inner assertions won't run because the exception will be raised before they do. You'll need to use `django.test.ignore_warnings()` to test the functionality. Also, it's a good idea to use `assertRaisesMessage` to check the message as well.

**Comment by timgraham** [comment]
Please see [Deprecating a feature](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature) for some documentation requirements that are missing.

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: django/utils/functional.py (hunk 1 above)
Comment (aaugustin) [inline django/utils/functional.py]: Yes, `django/utils/deprecation.py` would be more appropriate. We can tweak the message when we find new use cases, if any.

File: docs/ref/contrib/auth.txt [hunk 3]
```diff
@@ -111,39 +111,56 @@ Fields
...
 
-        Always returns ``False``. This is a way of differentiating
-        :class:`~django.contrib.auth.models.User` and
+        Read-only attribute which is always ``False``. This is a way of
+        differentiating :class:`~django.contrib.auth.models.User` and
         :class:`~django.contrib.auth.models.AnonymousUser` objects.
         Generally, you should prefer using
-        :meth:`~django.contrib.auth.models.User.is_authenticated()` to this
-        method.
+        :attr:`~django.contrib.auth.models.User.is_authenticated` to this
```
Comment (aaugustin) [inline docs/ref/contrib/auth.txt]: Since `is_authenticated` is the preferred variant, it should be documented before `is_anonymous`.

File: docs/topics/auth/customizing.txt [hunk 4]
```diff
@@ -612,20 +612,39 @@ The following methods are available on any subclass of
...
-        Always returns ``False``. This is a way of differentiating
-        from  :class:`~django.contrib.auth.models.AnonymousUser` objects.
+        Read-only attribute which is always ``False``. This is a way of
+        differentiating :class:`~django.contrib.auth.models.User` and
+        :class:`~django.contrib.auth.models.AnonymousUser` objects.
         Generally, you should prefer using
-        :meth:`~django.contrib.auth.models.AbstractBaseUser.is_authenticated()` to this
-        method.
+        :attr:`~django.contrib.auth.models.User.is_authenticated` to this
+        attribute.
```
Comment (aaugustin) [inline docs/topics/auth/customizing.txt]: As above, swap the order.

**Comment by aaugustin** [comment]
The patch looks good. Let's merge it once the comments above are addressed.

**Comment by jlaine** [comment]
I believe I have addressed all comments, but do please let me know if I missed something.

File: docs/releases/1.10.txt [hunk 5]
```diff
@@ -832,6 +832,28 @@ features, is deprecated. Replace it with a custom lookup::
     models.CharField.register_lookup(Search)
     models.TextField.register_lookup(Search)
 
+Using ``request.user.is_authenticated()`` and ``request.user.is_anonymous()`` as methods
+----------------------------------------------------------------------------------------
+
+The following members of
+:class:`~django.contrib.auth.models.AbstractBaseUser` and
+:class:`~django.contrib.auth.models.AnonymousUser` classes are now
+properties, and should no longer be invoked as methods:
```
Comment (timgraham) [inline docs/releases/1.10.txt]: We should explain the rationale for the change to avoid criticism for useless deprecations.

**Comment by timgraham** [comment]
What about the case of custom user models overriding these methods? It should be documented as a backwards-incompatible change or a "compatibility" system check should be added to detect it and warn users.

File: docs/releases/1.10.txt (hunk 5 above)
Comment (jlaine) [inline docs/releases/1.10.txt]: OK I agree

**Comment by jlaine** [comment]
I am not sure how to move forrward on the "customer user models" topic..

File: docs/releases/1.10.txt [hunk 6]
```diff
@@ -838,7 +838,7 @@ Using ``request.user.is_authenticated()`` and ``request.user.is_anonymous()`` as
 The following members of
 :class:`~django.contrib.auth.models.AbstractBaseUser` and
 :class:`~django.contrib.auth.models.AnonymousUser` classes are now
-properties, and should no longer be invoked as methods:
+properties, and should no longer be invoked as methods::
```
Comment (uranusjr) [inline docs/releases/1.10.txt]: Is this change needed? Since the next block is not code, single colon seems more natural to me.

**Comment by timgraham** [comment]
I added some docs about the backwards-incompatible nature of the change for custom user models that override these methods. Merged in c1aec0feda73ede09503192a66f973598aef901d, thanks!
